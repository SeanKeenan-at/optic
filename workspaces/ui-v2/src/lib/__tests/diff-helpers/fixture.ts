import path from 'path';
import {
  buildUniverse,
  ITestUniverse,
} from '<src>/lib/__tests/diff-helpers/universes/buildUniverse';
import { ParsedDiff } from '<src>/lib/parse-diff';
import { DiffSet } from '<src>/lib/diff-set';
import { interpretShapeDiffs } from '<src>/lib/shape-diffs/shape-diffs';
import { IShapeTrail } from '@useoptic/cli-shared/build/diffs/shape-trail';
import colors from 'colors';
import {
  ICopy,
  ICopyStyle,
} from '<src>/optic-components/diffs/render/ICopyRender';
import { IChangeType, IInterpretation } from '<src>/lib/Interfaces';
import { IValueAffordanceSerializationWithCounter } from '@useoptic/cli-shared/build/diffs/initial-types';
import { newRegionInterpreters } from '<src>/lib/new-regions-interpreter';
import { IOpticDiffService } from '@useoptic/spectacle';

export const testCase = (basePath: string) => async (
  name: string
): Promise<{
  universe: ITestUniverse;
  diffSet: DiffSet;
  diffs: ParsedDiff[];
}> => {
  const universe = await buildUniverse(
    require(path.join(__dirname + '/universes', basePath, name + '.json'))
  );

  const started = await universe.opticContext.capturesService.startDiff(
    '123',
    'example-session'
  );
  await started.onComplete;

  const result = await universe.opticContext.diffRepository.findById('123');
  const diffs = (await result.listDiffs()).diffs;

  const parsedDiffs = diffs.map(
    ([diff, interactions, fingerprint]: any) =>
      new ParsedDiff(diff, interactions, fingerprint)
  );

  return {
    diffs: parsedDiffs,
    diffSet: new DiffSet(parsedDiffs, universe.currentSpecContext),
    universe,
  };
};
export const testCaseLoaded = async (
  example: any
): Promise<{
  universe: ITestUniverse;
  diffSet: DiffSet;
  diffs: ParsedDiff[];
}> => {
  const universe = await buildUniverse(example);

  const started = await universe.opticContext.capturesService.startDiff(
    '123',
    'example-session'
  );
  await started.onComplete;

  const result = await universe.opticContext.diffRepository.findById('123');
  const diffs = (await result.listDiffs()).diffs;

  const parsedDiffs = diffs.map(
    ([diff, interactions, fingerprint]: any) =>
      new ParsedDiff(diff, interactions, fingerprint)
  );

  return {
    diffs: parsedDiffs,
    diffSet: new DiffSet(parsedDiffs, universe.currentSpecContext),
    universe,
  };
};

export type IShapeDiffTestSnapshot = {
  preview: IInterpretation;
  commands: any[];
  trailValues: IValueAffordanceSerializationWithCounter;
};

export async function newRegionPreview(
  diff: ParsedDiff,
  diffService: IOpticDiffService,
  universe: ITestUniverse
) {
  const newRegion = await newRegionInterpreters(
    diff,
    diffService,
    universe.currentSpecContext
  );

  return newRegion;
}

export async function shapeDiffPreview(
  input: {
    shapeDiffGroupingHash: string;
    shapeTrail: IShapeTrail;
    diffs: ParsedDiff[];
  },
  universe: ITestUniverse
): Promise<IShapeDiffTestSnapshot> {
  const { pathId, method } = input.diffs[0].location(
    universe.currentSpecContext
  );

  const diffService = await universe.opticContext.diffRepository.findById(
    '123'
  );
  const shapeAffordances = await diffService.learnShapeDiffAffordances();

  const trailValues = shapeAffordances[input.diffs[0]!.diffHash];

  const preview = await interpretShapeDiffs(
    input.diffs[0].asShapeDiff(universe.currentSpecContext)!,
    trailValues,
    universe.spectacleQuery,
    universe.currentSpecContext
  );

  return {
    preview,
    trailValues,
    commands: [], //preview.toCommands(preview.updateSpecChoices!),
  };
}

export function ICopyToConsole(i: ICopy[]): string {
  return i
    .map((i) =>
      i.style === ICopyStyle.Code ? colors.bgBlue(i.text) : colors.green(i.text)
    )
    .join(' ');
}

export async function logResult(preview: any) {
  const toLog = [];

  toLog.push(
    `DIFF HASH: ${colors.underline(preview.diffDescription.diffHash)}`
  );

  toLog.push(
    `TITLE: ${colors.underline(
      ICopyToConsole(preview.overrideTitle || preview.diffDescription.title)
    )}`
  );

  toLog.push(
    `ASSERTION: ${colors.underline(
      ICopyToConsole(preview.diffDescription.assertion)
    )}`
  );
  toLog.push(
    `CHANGE TYPE: ${colors.underline(
      IChangeType[preview.diffDescription.changeType]
    )}`
  );

  toLog.push('CHOICES:');
  toLog.push('is field ' + preview.updateSpecChoices!.isField);
  toLog.push('is optional ' + preview.updateSpecChoices!.isOptional);
  toLog.push(
    preview
      .updateSpecChoices!.shapes.map(
        (i: any) => `   ${i.coreShapeKind}:${i.isValid}`
      )
      .join('  ')
  );

  console.log(preview.updateSpecChoices!.isField);

  console.log(toLog.join('\n'));
}
