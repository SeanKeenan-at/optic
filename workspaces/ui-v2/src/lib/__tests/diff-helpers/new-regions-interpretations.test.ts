import { newRegionPreview, testCase } from './fixture';

test('empty spec produced no body diffs', async () => {
  const universe = await testCase('empty-todo')('universe');
  const diffs = universe.diffSet.groupedByEndpointAndShapeTrail();
  expect(diffs.length).toBe(0);
});

test('spec with paths, and new bodies', async () => {
  const universe = await testCase('simple-todo')('partial-universe');
  const diffs = universe.diffSet.newRegions().groupedByEndpoint();

  expect(diffs).toMatchSnapshot();
});

test('new request body has correct commands', async () => {
  const universe = await testCase('simple-todo')('partial-universe');
  const diffs = universe.diffSet.newRegions().iterator();

  const diffService = await universe.universe.opticContext.diffRepository.findById(
    '123'
  );

  const preview = await newRegionPreview(
    diffs.find((i) => i.diffType === 'UnmatchedRequestBodyContentType')!,
    diffService,
    universe.universe
  );

  expect(preview).toMatchSnapshot();
});
//
test('new response body has correct commands', async () => {
  const universe = await testCase('simple-todo')('partial-universe');
  const diffs = universe.diffSet.newRegions().iterator();

  const diffService = await universe.universe.opticContext.diffRepository.findById(
    '123'
  );

  const preview = await newRegionPreview(
    diffs.find((i) => i.diffType === 'UnmatchedResponseBodyContentType')!,
    diffService,
    universe.universe
  );

  expect(preview).toMatchSnapshot();
});
