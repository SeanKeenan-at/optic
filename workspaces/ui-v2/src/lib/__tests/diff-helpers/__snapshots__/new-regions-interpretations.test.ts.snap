// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spec with paths, and new bodies 1`] = `Array []`;
